ddtrace/contrib/wsgi/wsgi.py

```diff
@@ -40,6 +42,44 @@
 )
 
 
+class _TracedIterable:
+    def __init__(self, resp_iter, span, parent_span):
+        self._close = getattr(resp_iter, "close", None)
+        if PY2:
+            self._next = getattr(iter(resp_iter), "next")
+        else:
+            self._next = getattr(iter(resp_iter), "__next__")
+
+        self.parent_span = parent_span
+        self.span = span
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        try:
+            return self._next()
+        except StopIteration:
+            self._close_spans()
+            raise
+        except Exception:
+            self.span.set_exc_info(*sys.exc_info())
+            self._close_spans()
+            raise
+
+    def next(self):
+        return self.__next__()
+
+    def close(self):
+        if self._close:
+            self._close()
+        self._close_spans()
+
+    def _close_spans(self):
+        self.span.finish()
```

Comment:
can we add a flag so that we don't finish the spans more than once? Also might be worth logging when a span is already finished since this would be unexpected behaviour.

technically the span api allows multiple finishes but we should ensure that it only happens once here since that is what we expect

Reply:
_TracedIterable._finish_spans() is called when `StopIteration` is raised and `Iterable.close()` is called. It's expected that this method will be called twice.

I added `_self_span_finished` to check if the spans were finished before calling `Span.finish()`. This should guard against finishing a span more than once.